### PR TITLE
reflection probe planar reflection for native

### PR DESCRIPTION
**Pull request by xubing0906**
Re: #

### Changelog

*

-------

### Continuous Integration

This pull request:

* [x] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

**Review by jk20012001 — APPROVED** [review]

File: cocos/render-scene/scene/reflection-probe.jsb.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
+/*
+ Copyright (c) 2021 Xiamen Yaji Software Co., Ltd.
```
Comment (minggo) [inline cocos/render-scene/scene/reflection-probe.jsb.ts]: error date

File: cocos/render-scene/scene/reflection-probe.jsb.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+ */
+ import { ClearFlagBit } from '../../gfx';
```
Comment (minggo) [inline cocos/render-scene/scene/reflection-probe.jsb.ts]: extra space

File: cocos/render-scene/scene/reflection-probe.jsb.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import { SKYBOX_FLAG } from './camera';
+ declare const jsb: any;
```
Comment (minggo) [inline cocos/render-scene/scene/reflection-probe.jsb.ts]: The same as above.

File: cocos/render-scene/scene/reflection-probe.jsb.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+ 
+ export enum ProbeClearFlag {
```
Comment (minggo) [inline cocos/render-scene/scene/reflection-probe.jsb.ts]: The same as above.

File: native/cocos/core/scene-graph/Scene.h [hunk 5]
```diff
@@ -52,6 +52,8 @@ class Scene final : public Node {
     void onBatchCreated(bool dontSyncChildPrefab) override;
     bool destroy() override;
 
+    int32_t getNewReflectionProbeId() { return _reflectionProbeId++; }
```
Comment (minggo) [inline native/cocos/core/scene-graph/Scene.h]: Please add const.

File: native/cocos/core/scene-graph/Scene.h [hunk 6]
```diff
@@ -64,6 +66,8 @@ class Scene final : public Node {
     IntrusivePtr<SceneGlobals> _globals;
     bool _inited{false};
 
+    int32_t _reflectionProbeId = 0;
```
Comment (minggo) [inline native/cocos/core/scene-graph/Scene.h]: `int32_t _reflectionProbeId{0}` as other codes

File: native/cocos/renderer/pipeline/PipelineUBO.h [hunk 7]
```diff
@@ -34,14 +34,15 @@ namespace scene {
...
+class RenderScene;
 } // namespace scene
 namespace pipeline {
 class RenderPipeline;
 class GlobalDSManager;
 class CC_DLL PipelineUBO final {
 public:
     static void updateGlobalUBOView(const scene::Camera *camera, ccstd::array<float, UBOGlobal::COUNT> *bufferView);
-    static void updateCameraUBOView(const RenderPipeline *pipeline, float *output, const scene::Camera *camera);
+    static void updateCameraUBOView(const RenderPipeline *pipeline, float *output, const scene::Camera *camera, const scene::RenderScene *renderScene = nullptr);
```
Comment (minggo) [inline native/cocos/renderer/pipeline/PipelineUBO.h]: It is better not use default value parameter.

File: native/cocos/renderer/pipeline/PipelineUBO.h [hunk 8]
```diff
@@ -54,7 +55,7 @@ class CC_DLL PipelineUBO final {
     void activate(gfx::Device *device, RenderPipeline *pipeline);
     void destroy();
     void updateGlobalUBO(const scene::Camera *camera);
-    void updateCameraUBO(const scene::Camera *camera);
+    void updateCameraUBO(const scene::Camera *camera, const scene::RenderScene *scene = nullptr);
```
Comment (minggo) [inline native/cocos/renderer/pipeline/PipelineUBO.h]: The same as above.

File: native/cocos/renderer/pipeline/ReflectionProbeBatchedQueue.cpp [hunk 9]
```diff
@@ -0,0 +1,195 @@
+/****************************************************************************
+ Copyright (c) 2020-2022 Xiamen Yaji Software Co., Ltd.
```
Comment (minggo) [inline native/cocos/renderer/pipeline/ReflectionProbeBatchedQueue.cpp]: error date

File: native/cocos/renderer/pipeline/ReflectionProbeBatchedQueue.h [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+class RenderBatchedQueue;
+class RenderPipeline;
+
+//const uint32_t phaseID(PassPhase::getPhaseID("shadow-caster"));
+
+class CC_DLL ReflectionProbeBatchedQueue final {
+public:
+    explicit ReflectionProbeBatchedQueue(RenderPipeline *);
+    ~ReflectionProbeBatchedQueue();
+    void destroy();
```
Comment (minggo) [inline native/cocos/renderer/pipeline/ReflectionProbeBatchedQueue.h]: Is destroy() needed? Isn't enough just use destructor?

File: native/cocos/renderer/pipeline/ReflectionProbeBatchedQueue.h [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+    // weak reference
+    ccstd::vector<scene::Pass *> _passes;
+    // weak reference
+    ccstd::vector<gfx::Shader *> _shaders;
+    // manage memory manually
+    RenderInstancedQueue *_instancedQueue{nullptr};
+    // manage memory manually
+    RenderBatchedQueue *_batchedQueue{nullptr};
+    uint32_t _phaseID = 0;
+    uint32_t _phaseReflectMapID = 0;
```
Comment (minggo) [inline native/cocos/renderer/pipeline/ReflectionProbeBatchedQueue.h]: Please use `{}` to initialize value.

File: native/cocos/renderer/pipeline/ReflectionProbeManager.cpp [hunk 12]
```diff
@@ -0,0 +1,55 @@
+/****************************************************************************
+ Copyright (c) 2020-2022 Xiamen Yaji Software Co., Ltd.
```
Comment (minggo) [inline native/cocos/renderer/pipeline/ReflectionProbeManager.cpp]: error date

File: native/cocos/renderer/pipeline/ReflectionProbeBatchedQueue.h [hunk 13]
```diff
@@ -0,0 +1,80 @@
+/****************************************************************************
+ Copyright (c) 2020-2022 Xiamen Yaji Software Co., Ltd.
```
Comment (minggo) [inline native/cocos/renderer/pipeline/ReflectionProbeBatchedQueue.h]: error date

File: native/cocos/renderer/pipeline/ReflectionProbeManager.cpp [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+
+ReflectionProbeManager::ReflectionProbeManager() = default;
+void ReflectionProbeManager::registerProbe(scene::ReflectionProbe* probe) {
+    _probes.emplace_back(probe);
+}
+void ReflectionProbeManager::unRegisterProbe(scene::ReflectionProbe* probe) {
+    for (auto* p : _probes) {
+        const auto iter = std::find(_probes.begin(), _probes.end(), probe);
+        _probes.erase(iter);
+        break;
```
Comment (minggo) [inline native/cocos/renderer/pipeline/ReflectionProbeManager.cpp]: Why need a loop?

File: native/cocos/scene/ReflectionProbe.cpp [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+#include "scene/ReflectionProbe.h"
+#include "Define.h"
+#include "core/scene-graph/Scene.h"
+#include "math/Quaternion.h"
+#include "renderer/pipeline/ReflectionProbeManager.h"
+namespace cc {
+namespace scene {
+ReflectionProbe::ReflectionProbe(int32_t id) {
+    _probeId = id;
+    pipeline::ReflectionProbeManager::getInstance()->registerProbe(this);
```
Comment (minggo) [inline native/cocos/scene/ReflectionProbe.cpp]: Why put ReflectionProbeManager in pipeline?

File: native/cocos/core/scene-graph/Scene.h (hunk 5 above)
Comment (xubing0906) [reply native/cocos/core/scene-graph/Scene.h]: _reflectionProbeId need to be ++

Comment (minggo) [reply native/cocos/core/scene-graph/Scene.h]: Why do it like this? It is wired.

**Review by star-e — APPROVED** [review]

File: native/cocos/core/scene-graph/Scene.h (hunk 5 above)
Comment (minggo) [reply native/cocos/core/scene-graph/Scene.h]: And please add inline for functions implemented in header file.

Comment (xubing0906) [reply native/cocos/core/scene-graph/Scene.h]: when add a reflection probe to a scene, new id needs to be generated, these codes sync ts engine

**Review by minggo — APPROVED** [review]

**Pull request merged**